IntraNodeComm primitives for allgather_matmul

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122163
* __->__ #118038



cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

Commit:
IntraNodeComm primitives for allgather_matmul

[ghstack-poisoned]

Commit:
Update on "IntraNodeComm primitives for allgather_matmul"


This PR introduces low-level primitives in `IntraNodeComm` for implementing efficient `allgather_matmul`.

[Looped CollectiveEinsum](https://dl.acm.org/doi/pdf/10.1145/3567955.3567959) proposed a technique that decomposes dependent matmuls and communication for more overlapping oppurtunity. The technique can greatly reduce the exposed communication in sequence parallel training and tensor parallel inference. We would like to enable torch.compile to automatically perform such optimizations.

The design of the primitives are driven by lw's implementation of the technique in xformers, and his crucial insights for achieving optimal performance with the technique on GPUs + NVLink/NVSwitch:
- Communication ops that uses SMs to perform copy from communication buffer to user buffer can greatly degrade the perf of concurrent matmuls. The current nccl kernels (including dev kernels) fall into this category. However, `cudaMemcpyAsync` on IPC buffers would perform "SM-free" copies via DMA engine.
- Decomposed matmuls suffer from wave quantization effect, which can be mitigated via running the matmuls in two concurrent streams.

Based on these insights, this PR introduces low-level primitives in `IntraNodeComm` for implementing efficient `allgather_matmul`, and an example/benchmark for `allgather_matmul`.

### Primitives
- `put()`/`get()` tensor into/from IPC buffers managed by `IntraNodeComm`. This allows `allgather_matmul` to be implemented with MPI-style one-sided communication which minimizes synchronization overhead on certain topologies (e.g. fully-connected with NVSwitch). The idea is inspired by minsii's recent exploration on accelerating TransformerEngine using the concept of shared window in the communication layer.
- `barrier()` - an IPC buffer based barrier that allows for synchronization among any subset of members.

### Example
Added an example implementation of `allgather_matmul` in `benchmarks/distributed/intra_node_comm/allgather_matmul.py`. It demonstrate that with the primitives, efficient `allgather_matmul` can be implemented with a small amount of Python code. The following benchmark is ran on an H100 host w/ NVSwitch:
```
allgather_matmul (M=2048, N=8192, K=2752, world_size=8): 1.365 ms/iter
matmul (M=16384, N=8192, K=2752): 1.342 ms/iter
```

<img width="566" alt="image" src="https://github.com/pytorch/pytorch/assets/4156752/1697f555-e54d-40b3-b3d0-34e2bf5cccce">

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

lw reviewed (commented):
Very nice work! I'm really impressed by how simple it is!

I have two high-level comments:

# Double copy
This approach requires two copies to send the data across rank: source tensor -> shared tensor -> target tensor. The version in xFormers manages to do only one copy, by making the matmul read directly from the shared tensor. I suspect this could have a perf impact, notably on how quickly the second matmul shard can begin.

# API design
I guess the reason we need two copies here is that the API structure is different: the comms is handled by a separate lower-level object which is not aware of the matmul, and this object doesn't want to expose access to its internal shared tensor.

I wonder if a higher-level more "opinionated" API would be more convenient and more efficient to use. IMO, we don't need the flexibility of send/recv, and instead we can implement all-gather and reduce-scatter collectives directly. The crucial difference with NCCL's one is that we need collectives _with signalling_, i.e., the collectives need to be able to notify completion of partial transfers which can then be waited-on by the matmul.

The API I'm imagining is something like this:
```
with comm.all_gather() as (shared_tensor, signal_counters):
    # We "acquire" the shared_tensor
    torch.matmul(my_shard, weight, out=out_shards[my_rank])
    for i in range(1, world_size):
        remote = (my_rank + i) % world_size
        wait_for_signal(signal_counters[remote])
        torch.matmul(shared_tensor[remote], weight, out=out_shards[remote])
    # By exiting the context manager we "release" the shared_tensor back to the comm
```

lw commented:
I also have some thoughts on the matmul side of these fused ops. For now you just implemented this as an example, and I thus don't know what your plans are. However, my two cents:
- The operators we have in xFormers are very CPU-heavy because they end up launching a lot of small kernels. This is problematic, and in general it would be better to be able to launch a fused kernel. We have some custom kernels in Triton, but they don't cover all cases.
- Specifically, in order to cover both the forward and backward passes of sequence parallelism, we need to handle a variety of combinations (communication on the M dimension, or on the K dimension, or two different matmuls fused with the same all-gather, ...). We found it impractical to cover them all with handcrafted Triton kernels, thus we believe that codegen is the more robust and scalable approach.
- Instead of reinventing codegen ourselves, I believe the best place to do this is by leveraging Inductor, thus having PyTorch compile a kernel on the fly that can wait for the signals and can do the matmul shards. Basically, it would automatically write the Triton kernels that we've been doing by hand so far.

I had been evoking this to @wanchaol as well, but as of now there's no plan/design for this.

yifuwang commented:
> This approach requires two copies to send the data across rank: source tensor -> shared tensor -> target tensor. The version in xFormers manages to do only one copy, by making the matmul read directly from the shared tensor. I suspect this could have a perf impact, notably on how quickly the second matmul shard can begin.

In this specific case, I don't think it has a perf impact, since the latency is approaching that of a non-decomposed matmul. Based on my interpretation of the trace, how quickly a matmul can begin is dictated by when the preceeding matmul reaches its final wave, rather than the completion of the preceding copy operation.

![image](https://github.com/pytorch/pytorch/assets/4156752/fe5883e7-7172-4dd3-a54a-eb81ee5a466b)

It indeed performs more copies. IIUC, the approach you described involves `world_size - 1` p2p copies per device. Whereas in the example, there are `world_size - 1` p2p copies plus one global-memory-to-ipc-buffer copy per device. However, the approach you described seems to require `world_size`x as much IPC buffer size (this would not be a problem if the compiler can utilize the buffer for other purposes).

I think for allgather-matmul, we can get away with smaller IPC buffer without exposing the extra copy. But for reduce-scatter-matmul, I think the approach you described is necessary for achieving optimal performance.

yifuwang commented:
> The operators we have in xFormers are very CPU-heavy because they end up launching a lot of small kernels. This is problematic, and in general it would be better to be able to launch a fused kernel. We have some custom kernels in Triton, but they don't cover all cases.

Curious if cuda graph is applicable in your use case (I want to understand how important it is to optimize for CPU overhead).

> Specifically, in order to cover both the forward and backward passes of sequence parallelism, we need to handle a variety of combinations (communication on the M dimension, or on the K dimension, or two different matmuls fused with the same all-gather, ...). We found it impractical to cover them all with handcrafted Triton kernels, thus we believe that codegen is the more robust and scalable approach.

Would love to learn more about these!

> Instead of reinventing codegen ourselves, I believe the best place to do this is by leveraging Inductor, thus having PyTorch compile a kernel on the fly that can wait for the signals and can do the matmul shards. Basically, it would automatically write the Triton kernels that we've been doing by hand so far.

Makes total sense.

wanchaol commented on benchmarks/distributed/intra_node_comm/allgather_matmul.py:
```diff
@@ -0,0 +1,128 @@
+    # buffer. After the barrier, all peers' input slices are accessible
+    # via their intra-node buffer without requiring synchronization.
+    with torch.cuda.stream(internal_stream):
+        comm.put(A_slice)
```
hmmm why we don't specify the offset when doing `put` or `get`? I think if we don't then all ranks would put its own slice to the beginning of the intra node buffer?

yifuwang replied:
> I think if we don't then all ranks would put its own slice to the beginning of the intra node buffer?

Yes this is intentional. What we are doing is essentially a "one-sided" broadcast where the sender puts data into its buffer and all receivers pull from it.

> hmmm why we don't specify the offset when doing put or get?

Imagine we do a scatter in a similar fashion (i.e. receiver pulled) where slices are produced in stages. At the beginning, we would put slice 0 at the beginning of the buffer. When slice 1 is ready, we could evict slice 0 and replace it with slice 1. However this requires additional synchronization with the consumer of slice 0. If the IPC buffer is abundant, instead, we could put slice 1 into the buffer at an offset.

wanchaol reviewed (commented):
This is great work! I have a few questions inlined and wondering why the offset was not used?

wanchaol commented:
> * Specifically, in order to cover both the forward and backward passes of sequence parallelism, we need to handle a variety of combinations (communication on the M dimension, or on the K dimension, or two different matmuls fused with the same all-gather, ...). We found it impractical to cover them all with handcrafted Triton kernels, thus we believe that codegen is the more robust and scalable approach.

@lw I thought for the transformer model, specifically transformerblock, we can mostly be sure that the sequence parallelism's allgather and reduce_scatter would happen in seq dim (which can always be dim 0)? wondering when we would hit communication on a different dimension.

For multiple matmuls fused with same allgather, are we referring to the attention q/k/v case?

awgu commented on benchmarks/distributed/intra_node_comm/allgather_matmul.py:
```diff
@@ -0,0 +1,128 @@
+    out = torch.empty((M, N), dtype=torch.bfloat16, device="cuda")
+
+    stride = M // world_size
+    A_slice = A[rank * stride : (rank + 1) * stride]
```
I think that `.chunk()` will incur unnecessary `as_strided` calls (`world_size` many) instead of just 1 for the `rank`.

yifuwang replied:
Ha I haven't really considered what @awgu said but his point makes total sense :)

lw replied:
You could also consider `A_slice = A.unflatten(0, (world_size, -1))[rank]` which I believe should be cheaper? (Plus you could "cache" the call to unflatten)

yifuwang commented on torch/csrc/distributed/c10d/intra_node_comm.cu:
```diff
@@ -731,5 +731,73 @@ at::Tensor allReduce(
+void IntraNodeComm::put(const at::Tensor& tensor, int64_t offset) {
+  TORCH_CHECK(
+      tensor.is_non_overlapping_and_dense(),
+      "IntraNodeComm::put: tensor must be non-overlapping and dense");
```
Overlapping means multiple tensor elements refer to the same storage location. Example:
```
>>> x = torch.arange(4)
>>> torch.as_strided(x, size=(5, 4), stride=(0, 1))
tensor([[0, 1, 2, 3],
        [0, 1, 2, 3],
        [0, 1, 2, 3],
        [0, 1, 2, 3],
        [0, 1, 2, 3]])
```

wanchaol replied:
Oh I see, so this is actually storage aliasing?

yifuwang commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -2389,6 +2389,25 @@ options :class:`~torch.distributed.ProcessGroupNCCL.Options`).
+  py::class_<IntraNodeComm, c10::intrusive_ptr<IntraNodeComm>>(
+      module, "_IntraNodeComm")
+      .def_static(
+          "rendezvous",
```
Yes this is temporary. We need to figure out the UX. e.g.:
- If the user wants to use intra-node comm explicitly, how would they bootstrap it?
- If the compiler decides that it's beneficial to use intra-node comm but the user program doesn't use it, how would the compiler bootstrap it?

Right this is exposed as a "private" pybind API for development purposes.

yifuwang commented on benchmarks/distributed/intra_node_comm/allgather_matmul.py:
```diff
@@ -0,0 +1,128 @@
+
+    torch.cuda.set_device(local_rank)
+    store = c10d.FileStore(f"/tmp/{rdzv_id}")
+    os.environ["ENABLE_INTRA_NODE_COMM"] = "1"
```
We'll need to have a round of design for UX and compiler integration after we identify the key set of primitives.

Commit:
Update on "IntraNodeComm primitives for allgather_matmul"


This PR introduces low-level primitives in `IntraNodeComm` for implementing efficient `allgather_matmul`.

[Looped CollectiveEinsum](https://dl.acm.org/doi/pdf/10.1145/3567955.3567959) proposed a technique that decomposes dependent matmuls and communication for more overlapping oppurtunity. The technique can greatly reduce the exposed communication in sequence parallel training and tensor parallel inference. We would like to enable torch.compile to automatically perform such optimizations.

The design of the primitives are driven by lw's implementation of the technique in xformers, and his crucial insights for achieving optimal performance with the technique on GPUs + NVLink/NVSwitch:
- Communication ops that uses SMs to perform copy from communication buffer to user buffer can greatly degrade the perf of concurrent matmuls. The current nccl kernels (including dev kernels) fall into this category. However, `cudaMemcpyAsync` on IPC buffers would perform "SM-free" copies via DMA engine.
- Decomposed matmuls suffer from wave quantization effect, which can be mitigated via running the matmuls in two concurrent streams.

Based on these insights, this PR introduces low-level primitives in `IntraNodeComm` for implementing efficient `allgather_matmul`, and an example/benchmark for `allgather_matmul`.

### Primitives
- `put()`/`get()` tensor into/from IPC buffers managed by `IntraNodeComm`. This allows `allgather_matmul` to be implemented with MPI-style one-sided communication which minimizes synchronization overhead on certain topologies (e.g. fully-connected with NVSwitch). The idea is inspired by minsii's recent exploration on accelerating TransformerEngine using the concept of shared window in the communication layer.
- `barrier()` - an IPC buffer based barrier that allows for synchronization among any subset of members.

### Example
Added an example implementation of `allgather_matmul` in `benchmarks/distributed/intra_node_comm/allgather_matmul.py`. It demonstrate that with the primitives, efficient `allgather_matmul` can be implemented with a small amount of Python code. The following benchmark is ran on an H100 host w/ NVSwitch:
```
allgather_matmul (M=2048, N=8192, K=2752, world_size=8): 1.365 ms/iter
matmul (M=16384, N=8192, K=2752): 1.342 ms/iter
```

<img width="566" alt="image" src="https://github.com/pytorch/pytorch/assets/4156752/1697f555-e54d-40b3-b3d0-34e2bf5cccce">

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "IntraNodeComm primitives for allgather_matmul"


This PR introduces low-level primitives in `IntraNodeComm` for implementing efficient `allgather_matmul`.

[Looped CollectiveEinsum](https://dl.acm.org/doi/pdf/10.1145/3567955.3567959) proposed a technique that decomposes dependent matmuls and communication for more overlapping oppurtunity. The technique can greatly reduce the exposed communication in sequence parallel training and tensor parallel inference. We would like to enable torch.compile to automatically perform such optimizations.

The design of the primitives are driven by lw's implementation of the technique in xformers, and his crucial insights for achieving optimal performance with the technique on GPUs + NVLink/NVSwitch:
- Communication ops that uses SMs to perform copy from communication buffer to user buffer can greatly degrade the perf of concurrent matmuls. The current nccl kernels (including dev kernels) fall into this category. However, `cudaMemcpyAsync` on IPC buffers would perform "SM-free" copies via DMA engine.
- Decomposed matmuls suffer from wave quantization effect, which can be mitigated via running the matmuls in two concurrent streams.

Based on these insights, this PR introduces low-level primitives in `IntraNodeComm` for implementing efficient `allgather_matmul`, and an example/benchmark for `allgather_matmul`.

### Primitives
- `put()`/`get()` tensor into/from IPC buffers managed by `IntraNodeComm`. This allows `allgather_matmul` to be implemented with MPI-style one-sided communication which minimizes synchronization overhead on certain topologies (e.g. fully-connected with NVSwitch). The idea is inspired by minsii's recent exploration on accelerating TransformerEngine using the concept of shared window in the communication layer.
- `barrier()` - an IPC buffer based barrier that allows for synchronization among any subset of members.

### Example
Added an example implementation of `allgather_matmul` in `benchmarks/distributed/intra_node_comm/allgather_matmul.py`. It demonstrate that with the primitives, efficient `allgather_matmul` can be implemented with a small amount of Python code. The following benchmark is ran on an H100 host w/ NVSwitch:
```
allgather_matmul (M=2048, N=8192, K=2752, world_size=8): 1.365 ms/iter
matmul (M=16384, N=8192, K=2752): 1.342 ms/iter
```

<img width="566" alt="image" src="https://github.com/pytorch/pytorch/assets/4156752/1697f555-e54d-40b3-b3d0-34e2bf5cccce">

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wanchaol commented on torch/csrc/distributed/c10d/intra_node_comm.cpp:
```diff
@@ -299,6 +299,10 @@ c10::intrusive_ptr<IntraNodeComm> IntraNodeComm::rendezvous(
     return nullptr;
   }
 
+  if (bufferSize == 0) {
```
🤔 what does this mean when passing in a `buffSize == 0`? I think the user intention on this might be they want to disable intra node comm?

yifuwang replied:
Yeah this is confusing. 0 here means "unspecified and use default value". Let me change it to `c10::optional` or something.

wanchaol commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -2404,6 +2404,26 @@ options :class:`~torch.distributed.ProcessGroupNCCL.Options`).
+          py::arg("rank"),
+          py::arg("world_size"),
+          py::arg("buffer_size") = 0)
+      .def("barrier", &IntraNodeComm::barrier, py::arg("ranks") = py::none())
```
it seems that the `ranks` are not used anywhere? do you for see that there would be cases that we only `barrier` in certain set of ranks? iiuc maybe we should always barrier on all ranks for this intra node communicator to align with ProcessGroup::barrier?

yifuwang replied:
> it seems that the ranks are not used anywhere?

This is used. The barrier in this PR supports synchronization among a subset of members.

> do you for see that there would be cases that we only barrier in certain set of ranks?

For example, to achieve efficient all-gather matmul on a host without NVSwitch but with a ring topology, a user can leverage finer-grained synchronization to create their own protocol.

> iiuc maybe we should always barrier on all ranks for this intra node communicator to align with ProcessGroup::barrier?

This will be the behavior when we expose this as `ProcessGroup::broad`. The finer-grained barrier will be exposed via different interfaces.

wanchaol commented on torch/csrc/distributed/c10d/intra_node_comm.cu:
```diff
@@ -698,5 +698,70 @@ int64_t getIntraNodeCommUsageCounter() {
+void IntraNodeComm::barrier(c10::optional<std::vector<int64_t>> ranks) {
+  if (!ranks.has_value()) {
+    ranks = std::vector<int64_t>(worldSize_);
+    std::iota(ranks->begin(), ranks->end(), 0);
```
hmmm would `IntraNodeComm` always have `0 - world_size` ranks?  and those ranks are always "local ranks" instead of global rank?

For a 2-D device mesh topology, the intra node ranks are not always from 0, i.e. one intra node comm rank can be `[0, 1, 2, 3]`, another intra node comm rank can be `[4, 5, 6, 7]`

yifuwang replied:
You brought up an interesting point. I double checked the relevant semantics:
- From the perspective of a process group, the ranks are always within the context of itself
- From the perspective of `torch.distributed` APIs, the ranks are global ranks

For example, let's we have a subgroup formed by global rank `[4, 5, 6, 7]` and we'd like to broadcast from rank 4:
- With `group.broadcast`, you need to specify the source rank as 0
- With `dist.broadcast`, you need to specify the source rank as 4
- There is a translation layer in [distributed_c10d.py](https://github.com/pytorch/pytorch/blob/main/torch/distributed/distributed_c10d.py#L2114)

It's certainly a bit confusing, but it's something we'll want to respect. For the purpose of this PR, we should use group-local ranks.

wanchaol reviewed (approved):
This looks great! The intra node primitives are pretty neat! Have a few comments inlined, stamp to unblock

Commit:
Update on "IntraNodeComm primitives for allgather_matmul"


This PR introduces low-level primitives in `IntraNodeComm` for implementing efficient `allgather_matmul`.

[Looped CollectiveEinsum](https://dl.acm.org/doi/pdf/10.1145/3567955.3567959) proposed a technique that decomposes dependent matmuls and communication for more overlapping oppurtunity. The technique can greatly reduce the exposed communication in sequence parallel training and tensor parallel inference. We would like to enable torch.compile to automatically perform such optimizations.

The design of the primitives are driven by lw's implementation of the technique in xformers, and his crucial insights for achieving optimal performance with the technique on GPUs + NVLink/NVSwitch:
- Communication ops that uses SMs to perform copy from communication buffer to user buffer can greatly degrade the perf of concurrent matmuls. The current nccl kernels (including dev kernels) fall into this category. However, `cudaMemcpyAsync` on IPC buffers would perform "SM-free" copies via DMA engine.
- Decomposed matmuls suffer from wave quantization effect, which can be mitigated via running the matmuls in two concurrent streams.

Based on these insights, this PR introduces low-level primitives in `IntraNodeComm` for implementing efficient `allgather_matmul`, and an example/benchmark for `allgather_matmul`.

### Primitives
- `put()`/`get()` tensor into/from IPC buffers managed by `IntraNodeComm`. This allows `allgather_matmul` to be implemented with MPI-style one-sided communication which minimizes synchronization overhead on certain topologies (e.g. fully-connected with NVSwitch). The idea is inspired by minsii's recent exploration on accelerating TransformerEngine using the concept of shared window in the communication layer.
- `barrier()` - an IPC buffer based barrier that allows for synchronization among any subset of members.

### Example
Added an example implementation of `allgather_matmul` in `benchmarks/distributed/intra_node_comm/allgather_matmul.py`. It demonstrate that with the primitives, efficient `allgather_matmul` can be implemented with a small amount of Python code. The following benchmark is ran on an H100 host w/ NVSwitch:
```
allgather_matmul (M=2048, N=8192, K=2752, world_size=8): 1.365 ms/iter
matmul (M=16384, N=8192, K=2752): 1.342 ms/iter
```

<img width="566" alt="image" src="https://github.com/pytorch/pytorch/assets/4156752/1697f555-e54d-40b3-b3d0-34e2bf5cccce">

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "IntraNodeComm primitives for allgather_matmul"


This PR introduces low-level primitives in `IntraNodeComm` for implementing efficient `allgather_matmul`.

[Looped CollectiveEinsum](https://dl.acm.org/doi/pdf/10.1145/3567955.3567959) proposed a technique that decomposes dependent matmuls and communication for more overlapping oppurtunity. The technique can greatly reduce the exposed communication in sequence parallel training and tensor parallel inference. We would like to enable torch.compile to automatically perform such optimizations.

The design of the primitives are driven by lw's implementation of the technique in xformers, and his crucial insights for achieving optimal performance with the technique on GPUs + NVLink/NVSwitch:
- Communication ops that uses SMs to perform copy from communication buffer to user buffer can greatly degrade the perf of concurrent matmuls. The current nccl kernels (including dev kernels) fall into this category. However, `cudaMemcpyAsync` on IPC buffers would perform "SM-free" copies via DMA engine.
- Decomposed matmuls suffer from wave quantization effect, which can be mitigated via running the matmuls in two concurrent streams.

Based on these insights, this PR introduces low-level primitives in `IntraNodeComm` for implementing efficient `allgather_matmul`, and an example/benchmark for `allgather_matmul`.

### Primitives
- `put()`/`get()` tensor into/from IPC buffers managed by `IntraNodeComm`. This allows `allgather_matmul` to be implemented with MPI-style one-sided communication which minimizes synchronization overhead on certain topologies (e.g. fully-connected with NVSwitch). The idea is inspired by minsii's recent exploration on accelerating TransformerEngine using the concept of shared window in the communication layer.
- `barrier()` - an IPC buffer based barrier that allows for synchronization among any subset of members.

### Example
Added an example implementation of `allgather_matmul` in `benchmarks/distributed/intra_node_comm/allgather_matmul.py`. It demonstrate that with the primitives, efficient `allgather_matmul` can be implemented with a small amount of Python code. The following benchmark is ran on an H100 host w/ NVSwitch:
```
allgather_matmul (M=2048, N=8192, K=2752, world_size=8): 1.365 ms/iter
matmul (M=16384, N=8192, K=2752): 1.342 ms/iter
```

<img width="566" alt="image" src="https://github.com/pytorch/pytorch/assets/4156752/1697f555-e54d-40b3-b3d0-34e2bf5cccce">

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "IntraNodeComm primitives for allgather_matmul"




[ghstack-poisoned]

Commit:
Update on "IntraNodeComm primitives for allgather_matmul"




[ghstack-poisoned]

yifuwang commented:
@pytorchbot merge
